File: day01/main.go
package main

import (
	"flag"
	"fmt"
	"io/ioutil"
	"sort"
	"strconv"
	"strings"
)

type CaloriesList = [][]int

// go run ./main.go -file='input.txt'
func main() {
	filePath := flag.String("file", "", "path to your input")

	flag.Parse()

	if *filePath == "" {
		fmt.Println("Please provide your input")
		return
	}

	b, err := ioutil.ReadFile(*filePath)

	if err != nil {
		fmt.Println(err)
		return
	}

	s := string(b)

	part1(s)

	fmt.Println()

	part2(s)
}

func parseInput(input string) CaloriesList {
	result := CaloriesList{}

	for _, calories := range strings.Split(input, "\n\n") {

		itemsStr := strings.Split(calories, "\n")
		items := []int{}

		for _, itemStr := range itemsStr {
			itemInt, _ := strconv.Atoi(itemStr)
			items = append(items, itemInt)
		}

		result = append(result, items)
	}

	return result
}

func sortCalories(c CaloriesList) []int {
	list := []int{}

	for _, calories := range c {
		curSum := 0

		for _, c := range calories {
			curSum += c
		}

		list = append(list, curSum)
	}

	sort.Ints(list)
	return list
}

func part1(input string) {
	parsed := parseInput(input)

	sorted := sortCalories(parsed)

	fmt.Printf("Part1 answer is: %d", sorted[len(sorted)-1])
}

func part2(input string) {
	parsed := parseInput(input)

	sorted := sortCalories(parsed)
	l := len(sorted)

	fmt.Printf("Part2 answer is: %d", sorted[l-1]+sorted[l-2]+sorted[l-3])
}

File: day01/main.ts
import { readStrFromArgs } from "../utils.ts"
import { _ } from "../deps.ts"

type Sums = number[] // Calories that every elf carrying (desc by amount)

const input = await readStrFromArgs()

part1(input) // 69501
part2(input) // 202346

function part1(input: string) {
  console.log("part1: ", parseInput(input)[0])
}

function part2(input: string) {
  const parsed = parseInput(input)
  console.log("part2: ", parsed[0] + parsed[1] + parsed[2])
}

function parseInput(input: string): Sums {
  const sums: Sums = []

  input.split("\n\n").forEach((chunk) => {
    sums.push(_.sum(chunk.split("\n").map(Number)))
  })

  sums.sort((a, b) => b - a)

  return sums
}
